File: scala-client/src/main/scala/com/couchbase/client/scala/durability/Durability.scala
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.scala.durability

import java.util.Optional

import com.couchbase.client.core.annotation.SinceCouchbase
import com.couchbase.client.core.msg.kv.{DurabilityLevel => CoreLevel}
import com.couchbase.client.core.service.kv.Observe
import com.couchbase.client.core.service.kv.Observe.{ObservePersistTo, ObserveReplicateTo}

/** Writes in Couchbase are written to a single node, and from there the Couchbase Server will
  * take care of sending that mutation to any configured replicas.  This option provides
  * some control over ensuring the success of the mutation's replication.
  *
  * This provides flexibility to the application to decide, per-operation, whether it wants to prioritise speed or
  * safety.
  */
sealed trait Durability {

  private[scala] def toDurabilityLevel: Optional[CoreLevel] = {
    this match {
      case Durability.Majority                   => Optional.of(CoreLevel.MAJORITY)
      case Durability.PersistToMajority          => Optional.of(CoreLevel.PERSIST_TO_MAJORITY)
      case Durability.MajorityAndPersistToActive =>
        Optional.of(CoreLevel.MAJORITY_AND_PERSIST_TO_ACTIVE)
      case _ => Optional.empty()
    }
  }
}

object Durability {

  /** The SDK will return as soon as the first single node has the mutation in memory (but not necessarily persisted
    * to disk).
    */
  case object Disabled extends Durability

  /** The SDK will do a simple polling loop to wait for the mutation to be available on a number of replicas.
    *
    * Available for all Couchbase Server versions.  If using Couchbase Server 6.5, then prefer using the other
    * Durability options.
    *
    * @param replicateTo the number of replicas to wait for on which the mutation will be available in-memory
    * @param persistTo   the number of replicas to wait for on which the mutation has been written to storage
    */
  case class ClientVerified(
      replicateTo: ReplicateTo.Value,
      persistTo: PersistTo.Value = PersistTo.None
  ) extends Durability

  /** The server will ensure that the change is available in memory on the majority of configured replicas, before
    * returning success to the SDK.
    *
    * Only available in Couchbase Server 6.5 and above.
    */
  @SinceCouchbase("6.5")
  case object Majority extends Durability

  /** The server will ensure that the change is available in memory on the majority of
    * configured replicas, plus persisted to disk on the active node, before returning success
    * to the SDK.
    *
    * Only available in Couchbase Server 6.5 and above.
    */
  @SinceCouchbase("6.5")
  case object MajorityAndPersistToActive extends Durability

  /** The server will ensure that the change is both available in memory and persisted to disk
    * on the majority of configured replicas, before returning success to the SDK.
    *
    * Only available in Couchbase Server 6.5 and above.
    */
  @SinceCouchbase("6.5")
  case object PersistToMajority extends Durability

}

/** The number of replicas to wait for on which the mutation will be available in-memory */
object ReplicateTo extends Enumeration {
  val None, One, Two, Three = Value

  private[scala] def asCore(v: ReplicateTo.Value): Observe.ObserveReplicateTo = v match {
    case None  => ObserveReplicateTo.NONE
    case One   => ObserveReplicateTo.ONE
    case Two   => ObserveReplicateTo.TWO
    case Three => ObserveReplicateTo.THREE
  }

}

/** The number of replicas to wait for on which the mutation has been written to storage */
object PersistTo extends Enumeration {
  val None, Active, One, Two, Three, Four = Value

  private[scala] def asCore(v: PersistTo.Value): Observe.ObservePersistTo = v match {
    case None   => ObservePersistTo.NONE
    case Active => ObservePersistTo.ACTIVE
    case One    => ObservePersistTo.ONE
    case Two    => ObservePersistTo.TWO
    case Three  => ObservePersistTo.THREE
    case Four   => ObservePersistTo.FOUR
  }
}

File: kotlin-client/src/main/kotlin/com/couchbase/client/kotlin/kv/Durability.kt
package com.couchbase.client.kotlin.kv

import com.couchbase.client.core.annotation.SinceCouchbase
import com.couchbase.client.core.api.kv.CoreDurability
import com.couchbase.client.core.msg.kv.DurabilityLevel
import com.couchbase.client.core.msg.kv.DurabilityLevel.MAJORITY
import com.couchbase.client.core.msg.kv.DurabilityLevel.MAJORITY_AND_PERSIST_TO_ACTIVE
import com.couchbase.client.core.msg.kv.DurabilityLevel.NONE
import com.couchbase.client.core.msg.kv.DurabilityLevel.PERSIST_TO_MAJORITY
import com.couchbase.client.core.service.kv.Observe.ObservePersistTo
import com.couchbase.client.core.service.kv.Observe.ObserveReplicateTo
import com.couchbase.client.kotlin.annotations.VolatileCouchbaseApi

/**
 * Specifies the durability requirements for a mutation.
 */
public sealed class Durability {
    internal abstract fun isPersistent(): Boolean

    internal abstract fun toCore(): CoreDurability

    @Deprecated(
        level = DeprecationLevel.WARNING,
        replaceWith = ReplaceWith("Durability.None", "com.couchbase.client.kotlin.kv.Durability"),
        message = "For consistency with other options, 'Disabled' is renamed to 'None'.",
    )
    public object Disabled : Durability() {
        override fun isPersistent(): Boolean = false
        override fun toCore(): CoreDurability = CoreDurability.NONE
        override fun toString(): String = "Disabled"
    }

    public object None : Durability() {
        override fun isPersistent(): Boolean = false
        override fun toCore(): CoreDurability = CoreDurability.NONE
        override fun toString(): String = "None"
    }

    @SinceCouchbase("6.5")
    public data class Synchronous internal constructor(
        val level: DurabilityLevel,
    ) : Durability() {
        override fun isPersistent(): Boolean =
            level == MAJORITY_AND_PERSIST_TO_ACTIVE || level == PERSIST_TO_MAJORITY

        override fun toCore(): CoreDurability = CoreDurability.of(level)
    }

    public data class ClientVerified internal constructor(
        val persistTo: PersistTo,
        val replicateTo: ReplicateTo,
    ) : Durability() {
        override fun isPersistent(): Boolean = persistTo != PersistTo.NONE
        override fun toCore(): CoreDurability = CoreDurability.of(persistTo.coreHandle, replicateTo.coreHandle)
    }

    public companion object {
        /**
         * The SDK will report success as soon as the node hosting the
         * active partition has the mutation in memory (but not necessarily
         * persisted to disk).
         */
        @Deprecated(
            level = DeprecationLevel.WARNING,
            message = "For consistency with other options, 'disabled' is renamed to 'none'.",
            replaceWith = ReplaceWith("none()", "static com.couchbase.client.kotlin.kv.Durability.none"),
        )
        public fun disabled(): Durability = Disabled

        /**
         * The SDK will report success as soon as the node hosting the
         * active partition has the mutation in memory (but not necessarily
         * persisted to disk).
         */
        public fun none(): Durability = None

        /**
         * The client will poll the server until the specified durability
         * requirements are observed.
         *
         * This strategy is supported by all Couchbase Server versions,
         * but has drawbacks. When using Couchbase Server 6.5 or later,
         * the other durability options may be preferable.
         */
        public fun clientVerified(persistTo: PersistTo, replicateTo: ReplicateTo = ReplicateTo.NONE): Durability =
            if (replicateTo == ReplicateTo.NONE && persistTo == PersistTo.NONE) None
            else ClientVerified(persistTo, replicateTo)

        /**
         * The mutation must be replicated to (that is, held in the memory
         * allocated to the bucket on) a majority of the Data Service nodes.
         */
        @SinceCouchbase("6.5")
        public fun majority(): Durability = Synchronous(MAJORITY)

        /**
         * The mutation must be persisted to a majority of the Data Service nodes.
         */
        @SinceCouchbase("6.5")
        public fun persistToMajority(): Durability = Synchronous(PERSIST_TO_MAJORITY)

        /**
         * The mutation must be replicated to a majority of the Data Service nodes.
         *
         * Additionally, it must be persisted (that is, written and synchronised to disk)
         * on the node hosting the active partition (vBucket) for the data.
         */
        @SinceCouchbase("6.5")
        public fun majorityAndPersistToActive(): Durability = Synchronous(MAJORITY_AND_PERSIST_TO_ACTIVE)

        internal fun of(coreLevel: DurabilityLevel): Durability =
            if (coreLevel == NONE) none() else Synchronous(coreLevel)
    }
}

public enum class ReplicateTo(internal val coreHandle: ObserveReplicateTo) {
    NONE(ObserveReplicateTo.NONE),
    ONE(ObserveReplicateTo.ONE),
    TWO(ObserveReplicateTo.TWO),
    THREE(ObserveReplicateTo.THREE);

    public companion object {
        @VolatileCouchbaseApi
        public fun replicas(replicas: Int): ReplicateTo = when (replicas) {
            0 -> NONE
            1 -> ONE
            2 -> TWO
            3 -> THREE
            else -> throw IllegalArgumentException("Requested replica count $replicas not in range [0,3]")
        }
    }
}

public enum class PersistTo(internal val coreHandle: ObservePersistTo) {
    NONE(ObservePersistTo.NONE),
    ACTIVE(ObservePersistTo.ACTIVE),
    ONE(ObservePersistTo.ONE),
    TWO(ObservePersistTo.TWO),
    THREE(ObservePersistTo.THREE),
    FOUR(ObservePersistTo.FOUR);
}
